### DISCUSSION TITLE
NiceGUI Component Bazaar

### DISCUSSION THREAD

**Discussion by WolfgangFahl** [zauberzeug/nicegui#1618] [Ideas / Feature Requests]
🌟 Proposal: Introducing the NiceGUI Component Bazaar 🌟

🔍 Problem Statement:
While NiceGUI has seen rapid adoption and offers a robust set of features, there is an untapped potential in community contributions. Custom components developed by individual developers often remain isolated and are not easily discoverable by the wider user base.

🌐 Solution: The NiceGUI Component Bazaar:
Imagine a marketplace, a hub, where developers can share, discover, and integrate custom components built for NiceGUI. A bazaar where innovation thrives, collaboration is seamless, and the limits of what NiceGUI can do are pushed daily.
💡 Key Features:

 *   Community-Driven Growth: By having a centralized bazaar, we harness the collective creativity of our community, resulting in an ever-growing library of components.
 *   Quality Assurance: Components can be vetted, reviewed, and even ranked by the community, ensuring high standards.
 *   Easy Integration: With clear documentation and possibly a unified installation process, integrating new components can be as easy as a single click or command.

🌱 Benefits:

 *   For Developers: A platform to showcase their work, get feedback, collaborate, and also discover components that can accelerate their projects.
 *   For NiceGUI: An ever-evolving ecosystem that can attract even more developers, increase adoption, and solidify its position as a top GUI library.

🚀 Roadmap:

*    Phase 1: Gather feedback from the community on essential features and components.
*    Phase 2: Design and develop the initial version of the Component Bazaar.
*    Phase 3: Launch with a set of initial components and gather feedback for iterative improvements.

🤝 Let's transform how we think about NiceGUI components and create a vibrant marketplace that serves, celebrates, and accelerates our community's collective genius.

see
https://github.com/justpy-org/justpy/issues/372

**Comment by natankeddem** [comment]
What would be the workflow to integrate these components into my project? Is there something like a package manager (similar to pip)? Is it just a website that I copy code from and paste into my project? Maybe a NiceGUI website you run locally and can install the components for you utilizing web based resources (I like this option)?

**Comment by rodja** [comment]
Do you have examples of other projects which provide such a bazar?

**Comment by WolfgangFahl** [comment]
Some examples - commercial and non commercial (entry numbers as of 2023-09-14):
* https://de.wordpress.org/plugins/ (59.650 entries)
* https://extensions.joomla.org/
* https://www.drupal.org/project/project_module
* https://addons.mozilla.org/de/firefox/ 
* https://plugins.jetbrains.com/
* https://marketplace.visualstudio.com/VSCode
* https://frontwork.dev/component-marketplace
* https://www.componentsource.com/search/products/Borland%20Delphi?f%5B0%5D=at%3A914099&sort=downloads
 
Wordpress is a good example how the success of the platform is directly linked to the success of the availability of plugins for specific needs. 

Borland Delphi used to have a component store at the time it was having its heyday.

**Reply by falkoschindler** [reply]
These are certainly very valid examples.
Although Wordpress reminds me of a rather painful experience with web development, where you constantly have to hope that every extension you're using is supported by your Wordpress version. And if one of these components dies, good luck patching it or finding a replacement.
ROS is another example: https://index.ros.org/packages/. Its poor ecosystem for third-party packages was one reason why we stepped away from ROS and implemented [our own robot operating system](https://github.com/zauberzeug/rosys/).
Maybe this is an inherent problem of such a bazaar. But in my point of view it's far from easy to do it right and to make it a success - something with great value for production-ready web applications and not just a collection of thousands of half-baked and soon to be  outdated toy components.
So besides finding the right scope and implementation for the component distribution mechanism, maybe the hardest challenge will be finding the right interface, incentives, algorithms and community members for evaluation, review, moderation and so on.

**Comment by natankeddem** [comment]
Just an idea on implementation. 
- Have a website which lists components. 
- Each component has a unique name that is linked to a github repo in a database on the web. 
- When you want to actually implement it, there will be a `Bazaar` helper class (in `app` maybe?). This class would take a list of requested components you wish to utilize. 
- `Bazaar` will download (using the database) the components to the chosen path and then provide you with the class to implement them. 

```py
from nicegui import app, ui


class FlashyBanner:
    def __init__(self, title) -> None:
        ui.label(title)


class MenuDisplay:
    def __init__(self, items) -> None:
        ui.label("MENU")
        with ui.column():
            for key, value in items.items():
                ui.label(f"{key} is ${value}")


class ShoppingCart:
    def __init__(self, items) -> None:
        ui.label("CART")
        with ui.column():
            for key, value in items.items():
                with ui.row():
                    ui.label(f"{key} is ${value}")
                    ui.button("BUY")


class Bazaar:
    def __init__(self, comps, path="bazzar") -> None:
        self.comps = comps
        for comp in self.comps:
            setattr(self.__class__, comp, eval(comp))


b = Bazaar(comps=["FlashyBanner", "MenuDisplay", "ShoppingCart"])
b.FlashyBanner(title="Buy My Food!")
food = {"Fish": 5, "Cheese": 15}
b.MenuDisplay(food)
b.ShoppingCart(food)


ui.run()

```
In a final implementation `app.Bazaar` would utilize a remote database and the local file system to manage the components. The component classes would all be in modules located in discrete paths ie `bazaar/FlashyBanner/FlashyBanner.py` (contains `class FlashyBanner():`) which `app.Bazaar` would manage access to. The user wouldn't need to manually handle any files at all. `app.Bazaar` would determine if the component was missing or old and would automatically download/update it.

I am pretty sure this is viable but there might be some reason why python could not implement this possibly. Also, there is the matter of offline behaviour that would need to be determined. This sounded like an interesting implementation to me but I am not sure if anyone else feels it is worth pursuing.

**Reply by falkoschindler** [reply]
Interesting idea, @natankeddem! I like starting from the user's perspective and playing with a possible API.
I see some challenges though:
- If I understand correctly, class implementations like `FlashyBanner` will be generated during runtime. So there is no support for type hints and autocompletion by the IDE. Maybe there's a way to fix that once the component is downloaded, but at the moment I don't see how.
- If `Bazaar` automatically updates from GitHub repos, we'd probably need some kind of version specification to avoid getting every unstable update from main.

**Reply by natankeddem** [reply]
I think you have digested my intent here quite well.

Type hinting will be one of the challenges that might not be solved or require some additional thought/research.  In the worse case the user will need to import the paths to the modules\elements in any case so no worse than if they were manually downloading code and throwing it in their folder.

The goal with this implementation is to not stray to far from the existing `ui.element` syntax used in NiceGUI already while providing a path to what are essentially dynamic elements. Then sticking some "magic" in the middle to act as essentially a package manager.

**Comment by rodja** [comment]
The Bazar could also be queried/used via an CLI (see https://github.com/zauberzeug/nicegui/discussions/1462). That way it would be similar to pip. But then, why not use pip in the first place? The Bazar could simply be a list of additional packages which extend NiceGUIs functionality. Like https://github.com/ofenbach/CatDesign for example.

**Reply by WolfgangFahl** [reply]
see http://ngdemo.bitplan.com/docs
and
http://ngdemo.bitplan.com/solutions.yaml

Currently developers seem to mostly use github and pypi - there are only a few gitlab users at this time. The [nicegui search on pypi](https://pypi.org/search/?q=nicegui) reveals some 27 packages and on github the query retrievs >160.

**Comment by WolfgangFahl** [comment]
A first step would be simply tagging repositories and pip entries with "nicegui". Then at least we'd have a list of applicants. Then the next step is some kind of quality rating to be applied which could have objective aspects such as compatbility tests and code and repository quality as well as subjective ones (user stars ...)

**Comment by WolfgangFahl** [comment]
see also https://github.com/zauberzeug/nicegui/tree/main/examples and the [nicegui documentation](https://nicegui.io/documentation)

**Comment by retsyo** [comment]
There is [such a site](https://components.streamlit.app/) for [streamlit](https://streamlit.io/).

**Comment by WolfgangFahl** [comment]
With the source code at https://github.com/jrieke/components-hub

**Comment by WolfgangFahl** [comment]
see https://github.com/WolfgangFahl/nicegui_widgets/issues/50 for the start of some actual work on this

![grafik](https://github.com/WolfgangFahl/nicegui_widgets/assets/1336221/bb004cda-2c85-4c2a-8cba-5b99e2e1c5db)

```json
{
  "name": "niceguitools",
  "package": "niceguitools",
  "demo": null,
  "forum_post": null,
  "github": null,
  "pypi": null,
  "image_url": null,
  "stars": null,
  "github_description": "# niceguitools\r\n",
  "pypi_description": "A collection of tools for nicegui",
  "avatar": null,
  "search_text": null,
  "github_author": null,
  "pypi_author": null,
  "created_at": null,
  "downloads": null,
  "categories": [],
  "version": "0.0.1"
}
```

**Comment by WolfgangFahl** [comment]
Demo is now at http://ngdemo.bitplan.com/solutions
![grafik](https://github.com/zauberzeug/nicegui/assets/1336221/889121be-599a-49e6-8ea7-8d09acb66a9f)

**Comment by WolfgangFahl** [comment]
and improved to 
<img width="1869" alt="grafik" src="https://github.com/zauberzeug/nicegui/assets/1336221/f9e7e420-8bf5-41c4-a043-34859194eee1">

**Comment by WolfgangFahl** [comment]
# Technical Guide for NiceGUI Project Inclusion in the Solutions List

## Objective
This guide outlines the steps required for your NiceGUI project's automatic inclusion in the NiceGUI solutions list, as detailed in the provided source code.

## Automatic Inclusion Criteria
Your project's presence in [http://ngdemo.bitplan.com/solutions](http://ngdemo.bitplan.com/solutions) is determined by specific metadata from your GitHub and PyPI repositories.

### GitHub Requirements
1. **Topic Tagging**: Add topics to your repository (e.g., `**nicegui**`, `python`). nicegui is a mandatory topic to be included.
   - Guide: [Classifying your repository with topics](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/classifying-your-repository-with-topics)
2. **Repository URL**: Ensure your project's GitHub repository URL is included in the metadata.
3. **Star Count**: The number of stars can enhance the project's visibility, although it's not a strict requirement.

### PyPI Requirements
1. **Package Name**: Your package name on PyPI should clearly represent your project.
2. **Metadata**: Accurately complete the `pyproject.toml` or `setup.py` with your project's metadata.
   - Guide: [Packaging Python Projects](https://packaging.python.org/tutorials/packaging-projects/)
3. **Project URLs**: In your `pyproject.toml` or `setup.py`, include URLs for the source code (GitHub link), tracker (issue tracking), and documentation.
4. **Description**: Provide a detailed description on PyPI, which will be used as the project's summary.

### Project Detail Attributes
If you have set up your project as described above, the solution list update will effectively retrieve and display your project. Important attributes include:
- Name
- GitHub URL
- PyPI URL
- Description (from both GitHub and PyPI)
- Number of GitHub Stars
- Project's author/maintainer
- Creation Date (GitHub)

## Steps for Developers
1. **Tag Your GitHub Repository**: Apply relevant topics to your GitHub repository.
2. **Update Metadata**: Ensure your `pyproject.toml` or `setup.py` accurately reflects your project's details on PyPI.
3. **Maintain Your Repository**: Regularly update and maintain your GitHub repository and PyPI listing for continued visibility.

## Candidate Projects
To explore existing candidate projects for NiceGUI or to gauge the landscape, you can use the following search queries:
- GitHub Search by Topic: [NiceGUI Projects on GitHub](https://github.com/search?q=topic%3Anicegui&type=repositories)
- PyPI Search: [nicegui Projects on PyPI](https://pypi.org/search/?q=nicegui)
- General GitHub Search: [NiceGUI Repositories](https://github.com/search?q=nicegui&type=repositories)

## Updating Project Information
Changes in your GitHub or PyPI metadata will be reflected in the NiceGUI solutions list during the next update cycle.

## Further Information
For more discussions or questions about the inclusion process, refer to the initial discussion at [NiceGUI GitHub Discussion #1618](https://github.com/zauberzeug/nicegui/discussions/1618).

**Comment by WolfgangFahl** [comment]
From the solutions/projects view I'd like now to go the the components. I think we should have 

- declaration
- usage
- demo
- installation

as the releavant parts. To pick up this information and standarize it i propose to add .components.yaml files to the project in which the components are listed with these details. Given that similar work has already been done for justpy and as a basis for justpy users wanting to migrate justpy components will be included in the solution lists with the hope to be able to link traditional justpy components to nicegui components.

**Comment by WolfgangFahl** [comment]
For the automated inclusion of components in the NiceGUI solutions list, a structured `components.yaml` file is essential. Here's an example from the [WolfgangFahl's nicegui_widgets](https://github.com/WolfgangFahl/nicegui_widgets) project:

see http://ngdemo.bitplan.com/components/WolfgangFahl_nicegui_widgets 
![grafik](https://github.com/zauberzeug/nicegui/assets/1336221/9ad97845-5a7c-4fe0-8d3d-a5688be29c39)

and the declaration at
https://github.com/WolfgangFahl/nicegui_widgets/blob/main/.components.yaml
```yaml
# Components
# WF 2023-12-16
components:
- name: dict_edit.py
  source: /ngwidgets/dict_edit.py
  issue: 47
  demo_url: http://ngdemo.bitplan.com/dictedit
  demo_image_url: https://github.com/WolfgangFahl/nicegui_widgets/assets/1336221/d1682fce-975a-4ecc-a82e-c2c397e8222a
- name: color_schema.py
  source: /ngwidgets/color_schema.py
  issue: 15
  demo_url: http://ngdemo.bitplan.com/color_schema
  demo_image_url: https://github.com/WolfgangFahl/nicegui_widgets/assets/1336221/2bb926d4-da56-4201-a04c-af032d4de08e
```

### Explanation of Fields in `components.yaml`

The `components.yaml` file serves as a catalog of components within a project. Each entry in this file describes a specific component with the following fields:

- `name`: 
  - **Description**: The name of the component file, typically a Python script.
  - **Example**: In `name: dict_edit.py`, `dict_edit.py` is the Python script containing the component's code.

- `source`: 
  - **Description**: The relative path to the component's source file within the project repository.
  - **Example**: `/ngwidgets/dict_edit.py` points to the location of `dict_edit.py` within the repository.

- `issue`: 
  - **Description**: A reference to the issue number in the project's issue tracker that relates to this component.
  - **Example**: `issue: 47` refers to issue number 47 in the project's GitHub issue tracker.

- `demo_url`: 
  - **Description**: A URL pointing to a live demo of the component.
  - **Example**: `http://ngdemo.bitplan.com/dictedit` provides a link to interact with the `dict_edit.py` component.

- `demo_image_url`: 
  - **Description**: A URL to an image showcasing the component. This helps users get a visual understanding of what the component looks like or how it functions.
  - **Example**: `https://github.com/WolfgangFahl/nicegui_widgets/assets/1336221/d1682fce-975a-4ecc-a82e-c2c397e8222a` is a link to an image of the `dict_edit.py` component.

**Comment by WolfgangFahl** [comment]
see https://www.reddit.com/r/nicegui/comments/18ke09e/nicegui_solutions_bazaar_prototype/

**Comment by retsyo** [comment]
well done, amigo

**Comment by patrickwasp** [comment]
maybe we can take inspiration from shadcn/ui? https://www.shadcn-vue.com/docs/components/accordion.html

**Comment by Alyxion** [comment]
First of all thanks for starting the discussion and your first catalog implementation approach @WolfgangFahl .

I also just wanted to propose shadn as well here but @patrickwasp was a bit faster. Not having worked with Wordpress yet like @falkoschindler but massively with Tensorflow, Keras, Pytorch etc. and all the extensions to those it became really a pain in the ass very quickly if you have an awesome library some student started at some point during his off-time ... you build your project upon it, the student does not care anymore and with every major change or update of tf/pytorch/keras and/or CUDA things get worse and worse, **especially** if he pushes it to pypi, includes some hard restrictions such as `python<=3.11`, requires some weird library only he and nobody else uses which is already deprecated since years etc. etc.

If you would on the other hand solve it the "shadn/ui way" like so:

```
nicegui bazar register https://github.com/WolfgangFahl/nicegui_widgets
```

Which scans the repo for a config file (e.g. official compatible NiceGUI versions w/ the option for force/ignore it, if you are sure it works nevertheless).

```
nicegui bazar add progress
```
... which then resolves to nicegui_widgets.progress if there is no collision or explicit via `nicegui bazar add nicegui_widgets.progress`.

Would then download the **full source code** but only the two files of that one single component (progress.py+progress.js) and all requirements (base components), e.g into a local "components" directory.

Within this component directory an config file would be maintained by the system, components.toml, yaml, what ever, a bit like pyproject.toml... which tracks which component had which absolute origin source, potentially checksum to be able to detect if the user made changes to the code etc. and to prevent them being overwritten by updates.

And if there was a major some level API change to NiceGUI which bricked one of your components... and the **student** is still alive:
```
nicegui bazar update *
```

And if the student is not alive anymore and it was just a minor API change, then:
* You could just fix it yourself, you have the full source locally after all
* If it was a popular component use something like ```nicegui bazar replace nicegui_widgets.progress still_alive_guy.progress```

All this if course requires well defined directory structures, unique names to prevent collisions, well isolated components so you don't have to embed like half the repo just to add a more beautiful button.

**Comment by WolfgangFahl** [comment]
@Alyxion https://github.com/WolfgangFahl/nicegui_widgets/issues/52 has the current description of how to add components. I have tried to get some open source projects to add a .component.yaml file as you can see with the linked tickets but so far the reactions were subpar. I doubt that we'll get much active support by developers (yet). For the time being we migh need to seek som AI approaches that do support the creation of component descriptions and checking for compliance with some rudimentary standards we might want to set. At least the current prototype shows candidate projects ...
